src/allmydata/scripts/tahoe_check.py

```diff
@@ -54,8 +55,12 @@ def check_location(options, where):
         return 1
     jdata = resp.read()
     if options.get("raw"):
-        stdout.write(jdata)
-        stdout.write("\n")
+        if PY3:
+            stdoutb = stdout.buffer
+        else:
+            stdoutb = stdout
+        stdoutb.write(jdata)
+        stdoutb.write(b"\n")
```

Comment:
I can see the argument for that. On the other hand, the "raw" option is specifically for passing through data unchanged. So I think maybe it's better not to mess with the contents at all? I guess the issue is when I/O encoding is different than network encoding. 

What do other reviewers think?

Reply:
Some observations:

* Itamar's version doesn't have a place for UnicodeDecodeError or UnicodeEncodeError to pop up since it gets bytes, doesn't transform them, and writes bytes
* This is a Tahoe-LAFS API and if it returns anything other than UTF-8 it is broken
* It's guaranteed 100% safe to write unicode to stdout on Python 3 because it *always* encodes using UTF-8 which can represent all codepoints (except I'm not exactly sure in which version of Python this became true)

So the extra decode/encode doesn't seem like much of a problem in practice.  I'm not exactly sure what the original intent of the "raw" option was.  It is documented as "Display raw JSON data instead of parsed.".  That's somewhat ambiguous to me.  I'm inclined to think this should really be "serialized" or something - since I think "parsed" in that description is really about JSON parsing not UTF-8 parsing.  So from that perspective I also don't mind having the `decode` happen.

I guess avoiding the `if PY3` is somewhat desirable and not touching `stdout.buffer` is probably better than touching it ... so yea I think I like jaraco's suggestion.

Reply:
Digging through the docs, the specific behavior on Unix (Windows is always UTF-8), is:

> the interpreter defaults to using the current locale settings, unless the current locale is identified as a legacy ASCII-based locale (as described for PYTHONCOERCECLOCALE), and locale coercion is either disabled or fails. In such legacy locales, the interpreter will default to enabling UTF-8 mode unless explicitly instructed not to do so.

This is only in Python 3.7, however.

```
$ LANG=en_US.US-ASCII LC_ALL=C python3
Python 3.6.13 (default, Feb 16 2021, 00:00:00) 
[GCC 10.2.1 20201125 (Red Hat 10.2.1-9)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import locale   
>>> locale.getpreferredencoding()
'ANSI_X3.4-1968'
>>> import sys
>>> sys.stdout.write("\u1234")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode character '\u1234' in position 0: ordinal not in range(128)
```

Moreover Python 3 won't always do UTF-8 on newer Python's, it might use some other encoding (though that seems unlikely in practice).

Some options:

* Drop Python 3.6, switch to 3.7. RHEL 8 has Python 3.8, Debian Stable has 3.7, but Ubuntu 18.04 is 3.6.
* Re-encode as UTF-8 regardless of locale, it's JSON, JSON is supposed to be UTF-8.
* Drop the `raw` feature.
* ...